runtime/service/src/main/java/org/apache/polaris/service/events/listeners/inmemory/InMemoryEventListener.java

```diff
@@ -0,0 +1,141 @@
+
+  @Context SecurityContext securityContext;
+  @Context ContainerRequestContext requestContext;
+
+  @VisibleForTesting
+  final LoadingCache<String, UnicastProcessor<PolarisEvent>> processors =
+      Caffeine.newBuilder()
+          .expireAfterAccess(Duration.ofHours(1))
+          .evictionListener(
+              (String realmId, UnicastProcessor<?> processor, RemovalCause cause) ->
+                  processor.onComplete())
+          .build(this::createProcessor);
+
+  @Override
+  protected void processEvent(PolarisEvent event) {
+    var realmId = callContext.getRealmContext().getRealmIdentifier();
+    processEvent(realmId, event);
+  }
+
+  protected void processEvent(String realmId, PolarisEvent event) {
+    var processor = Objects.requireNonNull(processors.get(realmId));
+    processor.onNext(event);
+  }
+
+  @Override
+  protected ContextSpecificInformation getContextSpecificInformation() {
+    var principal = securityContext.getUserPrincipal();
+    var principalName = principal == null ? null : principal.getName();
+    return new ContextSpecificInformation(clock.millis(), principalName);
+  }
+
+  @Nullable
+  @Override
+  protected String getRequestId() {
+    return (String) requestContext.getProperty(REQUEST_ID_KEY);
+  }
+
+  @PreDestroy
+  public void shutdown() {
+    processors.asMap().values().forEach(UnicastProcessor::onComplete);
+    processors.invalidateAll(); // doesn't call the eviction listener
+  }
+
+  protected UnicastProcessor<PolarisEvent> createProcessor(String realmId) {
+    UnicastProcessor<PolarisEvent> processor = UnicastProcessor.create();
+    processor
+        .emitOn(Infrastructure.getDefaultWorkerPool())
+        .group()
+        .intoLists()
+        .of(configuration.maxBufferSize(), configuration.bufferTime())
+        .subscribe()
+        .with(events -> flush(realmId, events), error -> onProcessorError(realmId, error));
+    return processor;
+  }
+
+  @Retry(maxRetries = 5, delay = 1000, jitter = 100)
+  @Fallback(fallbackMethod = "onFlushError")
+  protected void flush(String realmId, List<PolarisEvent> events) {
+    RealmContext realmContext = () -> realmId;
+    var metaStoreManager = metaStoreManagerFactory.getOrCreateMetaStoreManager(realmContext);
+    var basePersistence = metaStoreManagerFactory.getOrCreateSession(realmContext);
+    var callContext = new PolarisCallContext(realmContext, basePersistence);
+    metaStoreManager.writeEvents(callContext, events);
+  }
+
+  @SuppressWarnings("unused")
+  protected void onFlushError(String realmId, List<PolarisEvent> events, Throwable error) {
+    LOGGER.error("Failed to persist {} events for realm '{}'", events.size(), realmId, error);
+  }
+
+  protected void onProcessorError(String realmId, Throwable error) {
+    LOGGER.error(
+        "Unexpected error while processing events for realm '{}'; some events may have been dropped",
+        realmId,
+        error);
```

Comment:
I wonder if its just okay to drop events considering they are deemed to be Audit logs ?